test/studies/ml/lib/Tensor.chpl

```diff
@@ -0,0 +1,792 @@
+        var t = new Tensor({0..#(h + 2 * padding),0..#(w + 2 * padding),0..#c},real);
+        forall (i,j,c) in x.data.domain with (ref t) {
+            t[i + padding,j + padding,c] = x[i,j,c];
+        }
+        return t;
+    }
+
+    // Compute the resulting tensor shape of a cross correlation
+    proc correlateShape(filterShape: 2*int, inputShape: 2*int, stride: int, padding: int) {
+        const (kh,kw) = filterShape;
+        const (nh,nw) = inputShape;
+        if kh != kw then err("Correlation only works with square filters.", kh, " != ", kw);
+        return (AutoMath.floor((nh - kh + 2* padding):real / stride:real):int + 1, AutoMath.floor((nw - kw + 2 * padding):real / stride:real):int + 1);
+        // return (AutoMath.floor(((nh - kh + padding + stride):real) / (stride:real)):int,AutoMath.floor(((nw - kw + padding + stride):real) / (stride:real)):int);
+    }
+
+    // Compute the resulting tensor shape of a cross correlation
+    proc correlate(const ref filter: Tensor(?), const ref input: Tensor(?), stride: int = 1, padding: int = 0) {
+        if padding > 0 then 
+            return correlate_(filter,pad(input,padding),stride,padding);
+        return correlate_(filter=filter,input=input,stride,padding);
+    }
+
+    // Compute the resulting matrix of a cross correlation
+    proc correlate_(const ref filter: Tensor(2), const ref input: Tensor(2), stride: int, padding: int): Tensor(2) {
+        const (kh,kw) = filter.shape;
+        const (nh,nw) = input.shape;
+        if kh != kw then err("Correlation only works with square filters.", kh, " != ", kw);
+
+        const (outH,outW): 2*int = correlateShape((kh,kw),(nh,nw),stride,padding);
+        var corr = new Tensor({0..#outH,0..#outW},real);
+
+        forall (x,y) in corr.data.domain with (ref corr) {
+            var sum = 0.0;
+            forall (i,j) in filter.data.domain with (+ reduce sum) {
+                sum += input[x * stride + i, y * stride + j] * filter[i,j];
+            }
+            corr[x,y] = sum;
+        }
+        return corr;
+    }
+
+    // Compute the sum of cross correlations for each filter and input channel
+    proc correlate_(const ref filter: Tensor(3), const ref input: Tensor(3), stride: int, padding: int): Tensor(2) {
+        const (kh,kw,cIn) = filter.shape;
+        const (nh,nw,nc) = input.shape;
+        if kh != kw then err("Correlation only works with square filters.", kh, " != ", kw);
+        if cIn != nc then err("Correlation only works with filters and inputs of the same depth.", cIn, " != ", nc);
+
+        // const (outH,outW): 2*int = ((nh - kh + padding + stride) / stride,(nw - kw + padding + stride) / stride);
+        const (outH,outW): 2*int = correlateShape((kh,kw),(nh,nw),stride,padding);
+
+        var corr = new Tensor({0..#outH,0..#outW},real);
+
+        forall (x,y) in corr.data.domain with (ref corr) {
+            var sum = 0.0;
+            forall (i,j,c) in filter.data.domain with (+ reduce sum) {
```

Comment:
Another nested forall that can hurt performance, no need to change it just for the sake of it, though.